feat: implement automated migration process scaffold (#70)
COVESA/s2dm: .migration/docs/setup-guide.md

# S2DM Migration Workflow Setup Guide

This guide explains how to set up the automated S2DM migration workflow in your repository. The workflow automatically generates artifacts (GraphQL schemas, JSON schemas, registry files, and SKOS RDF) when changes are made to your specification files.

## Prerequisites

> - A repository with S2DM specification files in a `spec/` directory

Reviewer: ```suggestion
- A clean repository with a `spec/` directory containing schema files written in `GraphQL SDL` and following the data modeling ruleset of the `S2DM` approach.
> !NOTE If the repository is empty, or if the user is starting from scratch without any schema files, then the tools will create an empty `spec/` directory.
```

> - A `units.yaml` file at the repository root

Reviewer: This is wrong. `S2DM` is the one that provides predefined commonly used elements. Among them, [the unit_enums.graphql file](https://github.com/COVESA/s2dm/blob/main/src/s2dm/spec/unit_enums.graphql).
For now, this file is given as a reference that mirrows the [units from vss](https://github.com/COVESA/vehicle_signal_specification/blob/master/spec/units.yaml). However, there is already the plan to adopt and reuse standard model for units to construct this units file automatically (see #43).

> - GitHub Actions enabled in your repository

Reviewer: Add link or instructions on how to do this. The user might not have an idea of it.

## Setup Steps

### 1. Copy Files from S2DM Repository

> Copy the required files from the [COVESA/s2dm](https://github.com/COVESA/s2dm) repository to your repository:

Reviewer: What files? Introduce the file names and their purpose before instructing to copy them. 
Example: 
The migration process offers X, Y, and Z... 
To achieve that and set your repository up, it offers templates to automate the process.
So, the user has to copy those templates in the clean repo where the domain data model will be hosted..
etc...

This is easily done with any of the following methods:

Reviewer: provide an image or text block that shows how the directory and files structrure in the new repo should look like.

```

This script will:

- Initialize a git repository if needed
- Create the `spec/` directory
- Download the required workflow and configuration files
- Handle existing files with confirmation prompts

#### Option B: Download files directly from GitHub

1. Download the workflow file:
   - Go to https://github.com/COVESA/s2dm/blob/main/.migration/workflows/migrate.yml
   - Click "Raw" and save the file to `.github/workflows/migrate.yml` in your repository

2. Download the version bump configuration:
   - Go to https://github.com/COVESA/s2dm/blob/main/.migration/.bumpversion.toml
   - Click "Raw" and save the file to `.bumpversion.toml` at your repository root

#### Option C: Using Git (if you have both repositories locally)

```bash
cp path/to/s2dm/.migration/workflows/migrate.yml .github/workflows/migrate.yml
cp path/to/s2dm/.migration/.bumpversion.toml .bumpversion.toml
```

#### Option D: Using curl

```bash
mkdir -p .github/workflows

curl -o .github/workflows/migrate.yml https://raw.githubusercontent.com/COVESA/s2dm/main/.migration/workflows/migrate.yml

curl -o .bumpversion.toml https://raw.githubusercontent.com/COVESA/s2dm/main/.migration/.bumpversion.toml

> ```

Reviewer: is there a test/check for the user to make sure things went well?

### 2. Configure Repository Variables

The workflow supports optional configuration through GitHub repository variables.

> Go to your repository → Settings → Secrets and variables → Actions → Variables tab, and add:

Reviewer: What do they mean? Provide link or further information to the reader to understand the benefit of configuring this optional step. What implies to use and not to use them?

- **`SHACL_MODEL_NAMESPACE`**: Namespace for data model (e.g., `https://example.ns/model#`)
- **`SHACL_MODEL_PREFIX`**: Prefix for data model (e.g., `model`)

> Note: All variables are optional. If not set, the commands will use their default values.

### 3. Repository Structure

Ensure your repository has the following structure:

```plain
your-repo/
├── .github/
│   └── workflows/
│       └── migrate.yml          # Copied from COVESA/s2dm
├── .bumpversion.toml            # Copied from COVESA/s2dm
├── spec/                        # Your specification files (GraphQL)
│   └── ...
├── units.yaml                   # Units file
└── README.md
```

## How the Workflow Works

The migration workflow is triggered on pushes to the `main` branch when files in the `spec/` directory are modified.

### Critical Workflow Steps

#### 1. Version Bump Detection

```yaml
- name: Check spec version bump
```

- Compares current spec with the previous release
- Determines if changes require a `major`, `minor`, `patch`, or no version bump
- Skips the entire workflow if no version bump is needed

#### 2. Schema Composition

```yaml
- name: Compose GraphQL schema
```

> - Combines all specification files from `spec/` into a single GraphQL schema

Reviewer: ```suggestion
- Composes a valid GraphQL SDL schema from:
  * All the specification files of your domain model located inside the `<your-repo>/spec/` directory.
  * All the used items from the [pre-defined elements in the `S2DM`](https://github.com/COVESA/s2dm/tree/main/src/s2dm/spec) repo.
```

- Output: `.artifacts/graphql/schema.graphql`

#### 3. JSON Schema Generation

```yaml
- name: Generate JSON schema
```

- Converts the GraphQL schema to JSON Schema format

> - Output: `.artifacts/jsonschema/schema.json`

Reviewer: To avoid confusion to the reader, use consistent terminology. Instead of using similar terms like generate, convert, traslate, etc. Stick to the one that is aligned to the command group. i.e., `export`

#### 4. Registry Management

```yaml
- name: Initialize registry  # (first release only)
- name: Update registry      # (subsequent releases)
```

- **Initialize**: Creates the first registry with concept URIs and spec history
- **Update**: Updates existing registry with new concepts and tracks changes
- Output: `.artifacts/registry/history`

#### 5. SHACL Generation

```yaml
- name: Generate SHACL
```

- Creates SHACL (Shapes Constraint Language) validation shapes from GraphQL schema

> - Output: `.artifacts/shacl/schema.shacl.ttl` (Turtle RDF format)

Reviewer: Be consistent in terminology --> `export`

#### 6. SKOS RDF Generation

```yaml

> - name: Generate SKOS RDF

Reviewer: same here

```

> - Creates semantic vocabulary from GraphQL schema

Reviewer: ```suggestion
- Parses the reference specification (in GraphQL SDL) and generates a set of RDF triples that use the `SKOS` vocabulary, which enables the arbitrary extension of a vocabulary in a decoupled manner.
```

- Output: `.artifacts/skos/schema.ttl` (Turtle RDF format)

#### 7. VSpec Generation

```yaml

> - name: Generate VSpec

Reviewer: ```suggestion
- name: Generate vspec
```

```

> - Converts GraphQL schema to Vehicle Signal Specification (VSpec) format

Reviewer: ```suggestion
- Exports the GraphQL schema in the Vehicle Signal Specification (`.vspec`) format.
```
